### DISCUSSION TITLE
Google sign-in returns null session in production

### DISCUSSION THREAD

**Discussion by wbcdev** [nextauthjs/next-auth#3210] [Help] [answered]
### Question 💬

Hi,

I'm running into much trouble in getting the Google provider sign-in from our example app to work in production.  
I'm currently deploying the nextjs server to a gcloud function.  When I attempt log-in the session returned is either empty or is not being handled correctly afterwards, it's never actually worked, as of now it just flips me back to the homepage while not signing in.

My production domain is hosted and everything seems alright on Goolge's side.  Nothing else needs to be approved there.

At first there was this error in Google cloud functions logger:
/api/auth/error?error=OAuthCallback 

and concurrently my app was showing the same thing in the browser and not sending me back to the homepage, but was showing 'try logging in with a different provider'.  

The logger in Google Cloud Platform would also give this:

textPayload: "https://next-auth.js.org/errors#state_error OAuthCallbackError: Invalid state returned from OAuth provider"

“code: 'ERR_HTTP_HEADERS_SENT'”

“https://next-auth.js.org/errors#oauth_callback_error Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client”

I don't understand that headers part.

But interestingly, upon moving to an earlier version of my code it now just returns a sign-in requester back to the homepage of the app while still saying 'not signed-in' and showing session as 'null' when console logged from the header.js component, no more OAuthCallbackError, but I suspect the same cause of the behavior.  Everything runs smoothly in dev environment on localhost and when building and starting there.  This led me to believe that it is just a discrepancy between setting the environment variables in production vs. having them set in env.local.  

However, I still believe they are setting correctly on the function in deployment as I've tried changing NEXTAUTH_URL on the function (from mysite.web.app to my custom domain name) and it will reflect a change in production, I've looked at the GOOGLE_ID and secret and they're being set properly (no extra spaces).  I've tried putting NEXT_AUTH in front of my envs in production as well.  I even tried setting them in nextjs.config.

The api callback settings on google cloud platform seem to be fine, after all, they are working perfectly on localhost.  It's just that once I go to Google sign-in for the production domain no session is being returned, or at least my header component thinks so, maddening, but it should pick it up fine with useSession just at it does on localhost, right?  

I don't know if I need to be setting something up with jwt and session callbacks, as you can see from my commented out code below, but those didn't make any difference either on localhost (they worked either way for my case of just signing in) or in production (received the same error with or without jwt callback code) and I am to believe that the example code should just pretty much work as-is in production, so I'm going with that.

Has anyone else had this problem?  And is there a certain consideration I need to have for setting environment variables in production with firebase at least?

I thought maybe because there was a 'TrustedScript' console error on the Google sign-in that that was the problem.  But I'm not setting a Content Security Policy in the app and I really don't think it should be blocking, as it doesn't block me when production building on localhost even when giving that 'TrustedScript' page error in the console.  Whereas, on the other hand, if I try to set a csp for the app in total, and fail at it, it totally does just block me outright.  And I really hope it's not the csp because that's an entirely different question, how on earth to set it?

Thank you so much for helping out with this.  I have literally tried everything I can think of. 

### How to reproduce ☕️

To reproduce:  build example app and deploy nextjs server to firebase google cloud function.  Host the site with firebase and have index.html redirect to server.js through "main": "server.js" in package.json, thus your hosted url will run the server and the app.  Edit function on cloud console to see env vars or deploy the env vars when deploying the function.

package.json
```json
"dependencies": {
    "@apollo/client": "^3.4.16",
    "crypto": "^1.0.1",
    "dompurify": "^2.3.3",
    "firebase": "^9.1.3",
    "firebase-admin": "^9.12.0",
    "firebase-functions": "^3.15.7",
    "isomorphic-dompurify": "^0.16.0",
    "nanoid": "^3.1.30",
    "next": "^11.0.0",
    "next-auth": "latest",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "sqlite3": "^5.0.2"
  },
```

-----------------------------------------------------------------------------------------------
server.js
```
const admin = require("firebase-admin");
const functions = require("firebase-functions");
const next = require("next");
const config = require("./next.config");

admin.initializeApp();

const dev = process.env.NODE_ENV !== "production";
const app = next({
	dev,
	// the absolute directory from the package.json file that initialises this module
	// IE: the absolute path from the root of the Cloud Function
	conf: config,
});
const handle = app.getRequestHandler();

const server = functions.https.onRequest((request, response) => {
	// log the page.js file or resource being requested
	console.log("File: " + request.originalUrl);
	return app.prepare().then(() => handle(request, response));
});

exports.nextjs = { server };
```

------------------------------------------------------------------------------------------------
firebase.json
```json
{
	"emulators": {
		"functions": {
			"port": 5001
		},
		"firestore": {
			"port": 8080
		},
		"hosting": {
			"port": 5000
		},
		"ui": {
			"enabled": true
		}
	},
	"functions": {
		"source": ".",
		"ignore": [
			"firebase.json",
			"firebase-debug.log",
			"**/.*",
			"**/node_modules/**",
			"components/**",
			"utils/**",
			"pages/**",
			"public/**",
			"firestore.rules",
			"readme.md"
		],
    "runtime": "nodejs16"
	},
	"hosting":
		{
			"site": "mysite",
			"public": "public/",
			"cleanUrls": true,
      "ignore": [
      "firebase.json",
      "**/.*",
      "**/node_modules/**"
    ],
			"rewrites": [
				{
					"source": "**",
					"function": "nextjs-server"
				}
			]
		}
}
```

------------------------------------------------------------------------------------------------
[...nextauth.js]
```js
import NextAuth from "next-auth"
import Providers from "next-auth/providers"

export default NextAuth({
  providers: [
    Providers.Google({
      clientId: process.env.GOOGLE_ID,
      clientSecret: process.env.GOOGLE_SECRET,
      authorizationUrl: 'https://accounts.google.com/o/oauth2/v2/auth?prompt=consent&access_type=offline&response_type=code',
      state: false,
    }),
  ],

  secret: process.env.SECRET,

  // jwt: {
  //   //encryption: true,
  //   secret: 
  //   signingKey: 
  //   encryptionKey: 
  // },

  // callbacks: {
  //   async signIn(user, account, profile, email, credentials) { 
  //     console.log("User Is");
  //     console.log(user);
  //     return true, user, account 
  //   },

  //   async jwt(token, user, account, profile, isNewUser) {
  //     if (account?.accessToken) {
  //         token.accessToken = account.accessToken;
  //         token.user = user
  //     }
  //     return token;
  //   },

  //   async session(session, user) { 
  //     console.log("SESSION CALLED");
  //     session.accessToken = user.accessToken;
  //     session.expires = user.expires;
  //     return session 
  //   },
  // },

   session: {
      jwt: true,
      maxAge: 30 * 24 * 60 * 60, // 30 days
   },
  theme: 'light',

  debug: false,
})
```
---------------------------------------------------------------------------------------------------------------------
header.js
```js
import React from 'react'
import Link from 'next/link'
import { csrfToken, signIn, signOut, useSession, getSession } from 'next-auth/client'
import styles from './header.module.css'

export default function Header () {
  const [ session, loading ] = useSession()
  console.log('Session coming through!')
  console.log(session)
  
  return (
    <header>
      <noscript>
        <style>{`.nojs-show { opacity: 1; top: 0; }`}</style>
      </noscript>
      <div className={styles.signedInStatus}>
        <p className={`nojs-show ${(!session && loading) ? styles.loading : styles.loaded}`}>
          {!session && <>
            <span className={styles.notSignedInText}>You are not signed in</span>
            <a
                href={`/api/auth/signin`}
                className={styles.buttonPrimary}
                onClick={(e) => {
                  e.preventDefault()
                  signIn('google')
                }}
              >
                Sign in
              </a>
          </>}
          {session && <>
            {session.user.image && <span style={{backgroundImage: `url(${session.user.image})` }} className={styles.avatar}/>}
            <span className={styles.signedInText}>
              <small>Signed in as</small><br/>
              <strong>{session.user.email || session.user.name}</strong>
              </span>
            <a
                href={`/api/auth/signout`}
                className={styles.button}
                onClick={(e) => {
                  e.preventDefault()
                  signOut()
                }}
              >
                Sign out
              </a>
          </>}
        </p>
      </div>
      <nav>
        <ul className={styles.navItems}>
          <li className={styles.navItem}><Link href="/"><a>Home</a></Link></li>
          <li className={styles.navItem}><Link href="/client"><a>Client</a></Link></li>
          <li className={styles.navItem}><Link href="/server"><a>Server</a></Link></li>
          <li className={styles.navItem}><Link href="/protected"><a>Protected</a></Link></li>
          <li className={styles.navItem}><Link href="/api-example"><a>API</a></Link></li>
          <li className={styles.navItem}><Link href="/clients"><a>Clients</a></Link></li>
        </ul>
      </nav>
    </header>
  )
}
```

--------------------------------------------------------------------------
Also, I noticed from the network tab in production that the session request was a GET request 
----------------------------
Headers: Request Method: GET
Status Code: 200 

Response:  {}
-------------------------------

where I thought I read somewhere that it must be a POST request, which I thought was being forced when calling  signIn('google').

I really want to use NextAuth in my project, and I appreciate any help you may offer. 

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by balazsorban44** [comment, marked as answer]
Could be related to https://firebase.google.com/docs/hosting/manage-cache#using_cookies

> When using Firebase Hosting together with Cloud Functions or Cloud Run, cookies are generally stripped from incoming requests. This is necessary to allow for efficient CDN cache behavior. Only the specially-named __session cookie is permitted to pass through to the execution of your app.

That would mean Google strips all the necessary cookies for a successful login with `next-auth`

https://github.com/nextauthjs/next-auth/blob/76bf524e8ec144162bcd0474adba250028f9c4db/src/core/lib/cookie.ts#L55-L106

**Reply by wbcdev** [reply]
Yes!  That's it!  Thank you so much.  I wouldn't have found it without you.  I can now log in with this code:

cookies: {
  sessionToken: {
    name: `__session`,
    options: {
      httpOnly: true,
      sameSite: 'lax',
      path: '/',
      secure: true
    }
  },
},

Although now the jwt shows as null in the api/examples/jwt or the api page, which is ok with me.  I just want to log in!  Thanks again.

**Reply by wbcdev** [reply]
Balazs,

I am trying to upgrade to ver 4 of next-auth.js.  Everything is working smoothly in ver 3.  There is an issue similar to the one I raised above with ver 4 in production, however.  And, after researching a bit, I have to admit that I don't know where to turn on this one.  I really like NextAuth and want to use it in my project, and I feel like moving to version 4 is important in staying up to date with all of the new features.  

The code I used to get the google session to be named __session in ver 3 does not seem to be setting properly with ver 4.  When inspecting the header now I see

cookie: __Secure-next-auth.state=.....; __Secure-next-auth.pkce.code_verifier=.....   whereas with ver 3 in production I would see somewhere in there a __session

Everything is working fine, deploying, etc., when I update to version 4 except for my sign-in functionality, which gives an oAuth error, with the session cookie never being set to __session.  Also, I'm setting the name of the session token in cookies, as you'll see below.  This is an advanced option, and I am certainly not very comfortable using this option since it specifies in the NextAuth docs that I shouldn't be.  Is there a better way for me to set this cookie name so that Google is happy?  

Here is my [nextauth].js as closely as I could mimick it from my version 3, but for ver 4.


import NextAuth from "next-auth"
import GoogleProvider from "next-auth/providers/google"

export default NextAuth({
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_ID,
      clientSecret: process.env.GOOGLE_SECRET,
      authorization: 'https://accounts.google.com/o/oauth2/v2/auth?prompt=consent&access_type=offline&response_type=code',
    }),
  ],

  secret: process.env.SECRET,

  cookies: {
    sessionToken: {
      name: `__session`,
      options: {
        httpOnly: true,
        sameSite: 'lax',
        path: '/',
        secure: true
      }
    },

    // callbackUrl: {
    // name: `__Secure-next-auth.callback-url`,
    //   options: {
    //     sameSite: 'lax',
    //     path: '/',
    //     secure: true
    //   }
    // },
    // csrfToken: {
    // name: `__Host-next-auth.csrf-token`,
    //   options: {
    //     httpOnly: true,
    //     sameSite: 'lax',
    //     path: '/',
    //     secure: true
    //   }
    // },
    // pkceCodeVerifier: {
    //   name: `${cookiePrefix}next-auth.pkce.code_verifier`,
    //   options: {
    //     httpOnly: true,
    //     sameSite: 'lax',
    //     path: '/',
    //     secure: useSecureCookies
    //   }
    // },
    // state: {
    //   name: `${cookiePrefix}next-auth.state`,
    //   options: {
    //     httpOnly: true,
    //     sameSite: "lax",
    //     path: "/",
    //     secure: useSecureCookies,
    //   },
    // },
  },

  session: {
    //
    strategy: "jwt",
    //
  },

  jwt: {
    //
  },

  pages: {
    //
  },

  callbacks: {
     async signIn({ user, account, profile, email, credentials }) { return true },
     async session({ session, token, user }) { 
      session.accessToken = token 
      return session 
    },
    // async jwt({ token, user, account, profile, isNewUser }) { return token }
  },

  events: {},

  theme: {
    colorScheme: "light",
  },

  debug: false,
})


I hate to bother you and I appreciate your time.  If you or anybody else needs more information from me, please let me know.  Thank you.  

William

**Comment by shivamragnar** [comment]
Hey @wbcdev @balazsorban44  
I think i am also getting the same `{}` for Google login. My Twitter login works smoothly. While debugging i doubted if it is returning this response from the `cache` but since Twitter works fine, hence it should not be a problem with the cache.
However I came across this discussion, but I am still not sure whether to include these `cookie options` in my code. I tried including though and I keep getting the `Undefined error` for `__session` variable. 
Could you share more explanation, or your code about what worked for you? I don't think this is included anywhere in the docs. I am using Version 3 for now. I appreciate any help from you guys.

**Reply by wbcdev** [reply]
Heloo,

It might depend on where you are hosting.  I was hosting through google and using cloud functions, which was stripping away all session cookies except the one labelled '__session'.  This was causing my Google sign-in to return a null session.  I also was using version 3 of NextAuth, with the code for [next_auth].js the same as shown in the original post in this discussion.  

I just added the following cookie specifier code to that [next_auth].js file and it now allows a log-in in production with version 3 of NextAuth when hosted on gcf.  

cookies: {
sessionToken: {
name: __session,
options: {
httpOnly: true,
sameSite: 'lax',
path: '/',
secure: true
}
},

However, as you can see from my  above post, upon upgrading to version 4 of NextAuth, this code no longer seems to work.

I'm not sure why you're getting 'undefined' for the __session, but maybe my experience will help you.
When checking the network request header in the browser tools and looking at session, __session should be listed as a cookie if you are hosting from cloud functions. 

I am going to try to host from Vercel because I want to use Nextjs' ISR feature, a very valuable feature, which currently is not available when hosting from Google.  I am thinking that this may solve my problem.

**Comment by Mrjuanblack** [comment]
Hello, I'm using the credentials provider to sign in, and using firebase to check if those credentials are stored in firebase auth. I was getting the same issue described here when testing on prod (deployed next js app on firebase hosting using the experimental web framework option).
As soon as I changed the cookie name to `__session` I was finally able to log in but I'm curious about this warning log I see on firebase functions:

> Decoding Firebase session cookie failed. Make sure you passed the entire string JWT which represents a session cookie. See https://firebase.google.com/docs/auth/admin/manage-cookies for details on how to retrieve a session cookie.

I'm still able to navigate my page as if I was logged in and use protected API endpoints, but I would like to fix whatever is making that warning appear.

My auth options:

```typescript
import NextAuth, { AuthOptions } from "next-auth"
import CredentialsProvider from "next-auth/providers/credentials";
import {signInWithEmailAndPassword} from 'firebase/auth';
import { GetAuth } from "../../../../firebase/firebaseWebInit";

export const authOptions: AuthOptions = {
  cookies:{
    sessionToken:{
      name:'__session',
      options:{
        httpOnly: true,
        sameSite: 'lax',
        path: '/',
        secure: true
      }
    }
  },
  // Configure one or more authentication providers
  session: {
    strategy: 'jwt',
    maxAge: 5 * 24 * 60 * 60 // 5 days
  },
  providers: [
    CredentialsProvider({
      credentials: {},
      async authorize(credentials): Promise<any> {
        return await signInWithEmailAndPassword(GetAuth(), (credentials as any).email || '', (credentials as any).password || '')
          .then(userCredential => {
            if (userCredential.user) {
              return userCredential.user;
            }
            return null;
          })
          .catch(error => (console.log(error)))
          .catch((error) => {
            const errorCode = error.code;
            const errorMessage = error.message;
            console.log(error);
          });
      }
    })
  ],
  callbacks:{
    async jwt({ token, user, session }) {
			return token;
		},
		async session({ session, token, user }) {
			return session;
		},
    async signIn({ user, account, profile, email, credentials }) {
      return true;
    },
    async redirect({ url, baseUrl }) {
      return baseUrl
    }
  }
}
export default NextAuth(authOptions)
```

**Comment by klee3** [comment]
Yeah Same happening with the Discord provider. It works perfectly fine in Development but in Production it returns null while using useSession()
